File: 1_Review/python/person.py

class Person:
    def __init__(self, name, age):
        self.name = name
        self.age = age

    def greet(self):
        print(f"Hello, my name is {self.name} and I am {self.age} years old.")


class Dog:
    def __init__(self, name, age):
        self.name = name
        self.age = age

    def greet(self):
        print(f"Wuf")

class Cat:
    def __init__(self, name, age):
        self.name = name
        self.age = age

    def greet(self):
        print(f"Meow")

File: 1_Review/julia/person.jl
module PersonModule

export Person, Dog, Cat, greet

# In Julia, custom types are usually structs.
# "mutable" means we can change fields after creation.
mutable struct Person
    name::String
    age::Int
end

mutable struct Dog
    name::String
    age::Int
end

mutable struct Cat
    name::String
    age::Int
end

# Functions live outside structs; multiple dispatch picks the right one.
function greet(p::Person)
    println("Hello, my name is $(p.name) and I am $(p.age) years old.")
end

greet(d::Dog) = println("Wuf")
greet(c::Cat) = println("Meow")

end
